tiledb/sm/query/writers/global_order_writer.cc

```diff
@@ -1210,74 +1331,110 @@ Status GlobalOrderWriter::prepare_full_tiles_var(
   return Status::Ok();
 }
 
-GlobalOrderWriter::GlobalWriteState* GlobalOrderWriter::get_global_state() {
-  return global_write_state_.get();
-}
-
-std::pair<Status, std::unordered_map<std::string, VFS::MultiPartUploadState>>
-GlobalOrderWriter::multipart_upload_state(bool client) {
-  if (client) {
-    return {Status::Ok(), global_write_state_->multipart_upload_state_};
+uint64_t GlobalOrderWriter::num_tiles_to_write(
+    uint64_t start,
+    uint64_t tile_num,
+    std::unordered_map<std::string, WriterTileVector>& tiles) {
+  // Cache variables to prevent map lookups.
+  const auto buf_names = buffer_names();
+  std::vector<bool> var_size;
+  std::vector<bool> nullable;
+  std::vector<WriterTileVector*> writer_tile_vectors;
+  var_size.reserve(buf_names.size());
+  nullable.reserve(buf_names.size());
+  writer_tile_vectors.reserve(buf_names.size());
+  for (auto& name : buf_names) {
+    var_size.emplace_back(array_schema_.var_size(name));
+    nullable.emplace_back(array_schema_.is_nullable(name));
+    writer_tile_vectors.emplace_back(&tiles[name]);
   }
 
-  auto meta = global_write_state_->frag_meta_;
-  std::unordered_map<std::string, VFS::MultiPartUploadState> result;
-
-  // TODO: to be refactored, there are multiple places in writers where
-  // we iterate over the internal fragment files manually
-  const auto buf_names = buffer_names();
-  for (const auto& name : buf_names) {
-    auto&& [st1, uri] = meta->uri(name);
-    RETURN_NOT_OK_TUPLE(st1, {});
+  // Make sure we don't write more than the desired fragment size.
+  uint64_t size = current_fragment_size_;
```

Comment:
As far as I can tell, the `size` variable isn't necessary. Unless I'm missing something, you can just use `current_fragment_size_` directly and remove the line on 1387.